### PR TITLE
Metadata API: Implement threshold verification

### LINKED ISSUE #1306: Implement verification by a threshold of keys
**Description of issue or feature request**:
A verification of signed metadata by a threshold of keys is needed both for client and repository. 

`Metadata` already has vanilla one-signature verify and could potentially be the right location for this implementation. See https://github.com/theupdateframework/tuf/pull/1060#issuecomment-660056376 for thoughts about that distinction.

### PR DESCRIPTION
**Pull request by jku**
#1417 is merged so this is good to go

The delegating Metadata (root or targets) verifies that the delegated metadata is signed by required threshold of keys for the delegated role.

This implementation raises UnsignedMetadataError if threshold is not met. Other possible exceptions are programming errors.

Fixes #1306

### CODE REVIEW

**Review by sechkova — APPROVED** [review]
Maybe I've become biased after looking at the code in the next-gen client for a while because I couldn't point out any issues :) 

I was searching for a way to avoid adding `role_name` as an argument of `verify_delegate` but I saw you already opened https://github.com/theupdateframework/tuf/issues/1425 about it so +1 from me on that.

File: tests/test_api.py [hunk 1]
```diff
@@ -352,6 +352,56 @@ def test_metadata_timestamp(self):
...
+        root.verify_delegate('targets', targets)
+        targets.verify_delegate('role1', role1)
+        role1.verify_delegate('role2', role2)
+
+        # only root and targets can verify delegates
+        with self.assertRaises(ValueError):
+            snapshot.verify_delegate('snapshot', snapshot)
+        # cannot verify with non-existing role
+        with self.assertRaises(ValueError):
+            root.verify_delegate('role1', role1)
```
Comment (MVrachev) [inline tests/test_api.py:365]: Nit: If you use a role you haven't used before in the test it will be easier to understand.
Even though it's logical you can't delegate trust from `role1` to `role1`.

**Review by MVrachev — APPROVED** [review]
I didn't found any issues either.
LGTM!

File: tuf/api/metadata.py [hunk 2]
```diff
@@ -250,59 +250,61 @@ def sign(
...
+            raise ValueError(f"No delegation found for {role_name}")
+
+        # verify that delegate is signed by required threshold of unique keys
+        signing_keys = set()
+        for keyid in role.keyids:
+            key = keys[keyid]
+            try:
+                key.verify_signature(delegate, signed_serializer)
+                # keyids are unique. Try to make sure the public keys are too
+                signing_keys.add(key.keyval["public"])
```
Comment (MVrachev) [inline tuf/api/metadata.py]: Maybe making sure that we have `keyval["public"]` uniqueness can be done instead in `Root` and `Targets`?

File: tests/test_api.py [hunk 3]
```diff
@@ -359,6 +352,56 @@ def test_metadata_timestamp(self):
...
+        # different delegation keys should fail verification
+        with self.assertRaises(tuf.exceptions.UnsignedMetadataError):
+            root.verify_delegate('timestamp', snapshot)
+
+        # Higher threshold should fail verification
+        root.signed.roles['snapshot'].threshold += 1
+        with self.assertRaises(tuf.exceptions.UnsignedMetadataError):
+            root.verify_delegate('snapshot', snapshot)
+
+        # TODO test higher thresholds
```
Comment (MVrachev) [inline tests/test_api.py]: Wondering why `higher thresholds` will make a difference?

Comment (jku) [reply tests/test_api.py]: it's just that we don't have a test here that actually uses a threshold of more than one key successfully. It should work but... it still feels like we're testing the edge case (1 key) when the real problem is N keys.

File: tuf/api/metadata.py (hunk 2 above)
Comment (jku) [reply tuf/api/metadata.py]: yeah possible but I'd say it's a separate issue: #1429

File: tests/test_api.py (hunk 3 above)
Comment (jku) [reply tests/test_api.py]: The fact that I did not add a test case but just the comment does point to a problem in our testing infra: we probably need to be able to produce new signed metadata more easily in the test itself: this is something we should put some effort into (we could use repository_tool... but I don't think we want to even write the metadata to a file, we just want to be able to generate it in memory and test it)

File: tests/test_api.py (hunk 1 above)
Comment (jku) [reply tests/test_api.py:365]: This tries to verify that root delegates to role1 (and it does not so ValueError). The comment could be better though
`# try verifying a role that is not delegated by the delegator `

Comment (MVrachev) [reply tests/test_api.py:365]: Yes, I didn't read it right that it's the root that's calling it.🤦
Still, I agree that the comment can be improved.

File: tuf/api/metadata.py [hunk 4]
```diff
@@ -250,59 +250,61 @@ def sign(
...
+            keys = self.signed.keys
+            role = self.signed.roles.get(role_name, None)
+        elif isinstance(self.signed, Targets):
+            if self.signed.delegations:
+                keys = self.signed.delegations.keys
+                # Assume role names are unique in delegations.roles: #1426
+                roles = self.signed.delegations.roles
+                role = next((r for r in roles if r.name == role_name), None)
+        else:
+            raise ValueError("Call is valid only on delegator metadata")
```
Comment (jku) [inline tuf/api/metadata.py]: from Trishank: "Why ValueError?"

Valid question as arguments are not the issue here (except if we consider "self" an argument, which it is in a way). 

In any case it's not terribly important as this is a programming error and should not be handled by anyone

Comment (jku) [reply tuf/api/metadata.py]: I left it a ValueError: I'm not going to create a custom error for something that isn't meant to be caught and can't find a better one in the builtin exceptions.

File: tuf/api/metadata.py [hunk 5]
```diff
@@ -256,6 +256,61 @@ def sign(
...
+        role = None
+        if isinstance(self.signed, Root):
+            keys = self.signed.keys
+            role = self.signed.roles.get(role_name)
+        elif isinstance(self.signed, Targets):
+            if self.signed.delegations:
+                keys = self.signed.delegations.keys
+                # Assume role names are unique in delegations.roles: #1426
+                roles = self.signed.delegations.roles
+                role = next((r for r in roles if r.name == role_name), None)
```
Comment (jku) [inline tuf/api/metadata.py]: Also from trishank: "What does this line do? Looks like magic"
My response was:
> it returns the first role in delegations with name role_name (or None if no match).
> 
> I agree it's not pretty and a bit too clever.
> I could do something like this instead:
> 
> ```
> role = None
> for delegated_role in roles:
>     if delegated_role.name == role_name:
>         role = delegated_role
>         break
> ```
> 
> but that's 5 lines of _initialize-then-modify_ antipattern: not the end of the world but not perfect either.
> 
> the real fix is probably #1426 though (make roles a OrderedDict). Then this code becomes readable and efficient:
> `role = roles.get(role_name)`

I think I'd leave it as is ... but if others think that's too clever I'm happy to change to the more verbose one

File: tests/test_api.py (hunk 3 above)
Comment (jku) [reply tests/test_api.py]: I think the right way forward here is issue #1444

**Comment by MVrachev** [comment]
In our discussion with @jku  in issue #1438 we agreed that we should make sure that the metadata is accepted even if it has a couple of invalid keys, but it accommodates the requirement placed by threshold.
I think I don't see a test for that. Do you plan to add one @jku?

**Comment by jku** [comment]
> In our discussion with @jku in issue #1438 we agreed that we should make sure that the metadata is accepted even if it has a couple of invalid keys, but it accommodates the requirement placed by threshold.
> I think I don't see a test for that. Do you plan to add one @jku?

I think there are two things that we should definitely test:
* test that Key.verify_signature() raises UnsignedMetadataError (or potentially a derived error) if keys are in any way invalid. This  should happen in #1435, I've added the tests that I think make sense there
* test that verify_delegate() handles UnsignedMetadataErrors coming from Key.verify_signature(): catching that error is fine as long as threshold is eventually reached: this should happen in this PR. It does look like there is no test case with failing signatures that should eventually succeed (because threshold is still reached). That would make sense to add (but it does not need to be about invalid keys in particular).

Soo, yes I guess I plan to add one now...

File: tuf/api/metadata.py (hunk 4 above)
Comment (sechkova) [reply tuf/api/metadata.py]: Could be `TypeError` since what `isinstance` checks is the type but is it much better than `ValueError` .. I guess not.

File: tests/test_api.py (hunk 3 above)
Comment (jku) [reply tests/test_api.py]: I did end up adding tests for threshold of 2.

**Comment by jku** [comment]
rebased on develop, added a commit to add a tests for 
* signature that fails to verify, but overall signatures reach threshold
* threshold that actually contains multiple signatures

This also changes one error type as suggested by teodora

File: tuf/api/metadata.py (hunk 4 above)
Comment (jku) [reply tuf/api/metadata.py]: I originally did not pick that because the object type the call is made on (Metadata) is correct ... but I think you're right, it's not perfect but it's better.

**Review by MVrachev — APPROVED** [review]
Thanks for adding the tests!
Seems good to me!

File: tuf/api/metadata.py (hunk 2 above)
Comment (jku) [reply tuf/api/metadata.py]: I only very recently realised the TUF spec does not require the existence of "public" at all -- the three keys defined in the spec all happen to have it but it isn't required.

So this line needs to change. The spec does require that KEYIDs are unique so `verify_delegate()` will just respect that and will not try to poke into the internals.

**Comment by jku** [comment]
Added a commit that removes the attempt to enforce key contents uniqueness:
* I only very recently realised the TUF spec does not require the existence of "public" key within the Key.keyval dictionary at all -- the three keys defined in the spec all happen to have it but it isn't required.
* We know that the attempt is not protection against maliciously crafted keys anyway: creating multiple representations of the same key is not difficult 

Luckily the spec requires that KEYIDs are unique so verify_delegate() will just respect that and will not try to poke into the internals of the keys.

File: tuf/api/metadata.py [hunk 6]
```diff
@@ -273,6 +276,60 @@ def sign(
...
+        role_name: str,
+        delegate: "Metadata",
+        signed_serializer: Optional[SignedSerializer] = None,
+    ):
+        """Verifies that 'delegate' is signed with the required threshold of
+        keys for the delegated role 'role_name'.
+
+        Args:
+            role_name: Name of the delegated role to verify
+            delegate: The Metadata object for the delegated role
```
Comment (joshuagl) [inline tuf/api/metadata.py]: My reading of this is that `verify_delegate` takes an identifier for the delegated role (role_name) and an instance of the delegated role (delegate) – I would expect these two named parameters to have similar names.
* `delegated_role` and `delegated_metadata`?
* `role_name` and `role_metadata`? (handily avoiding the delegate, delegated, delegatee, delegator confusion)

File: tuf/api/metadata.py (hunk 5 above)
Comment (joshuagl) [reply tuf/api/metadata.py]: Perhaps add a one-liner comment for less familiar readers?
> Find the first role in delegations with name role_name (or None if no match)

**Review by joshuagl — COMMENTED** [review]
This looks good to me. I wonder if the parameter names in `verify_delegate` could be renamed to more clearly indicate that role_name and delegate are related (an identifier and an instance)?

**Comment by jku** [comment]
I chose the more verbose names: it's easy to lose track of where the keys and thresholds really come from and what they are applied on so more verbose seems reasonable.

**Comment by jku** [comment]
Someone made mypy rules stricter in the mean time :grimacing: so had to rebase and fix the annotation (no changes in existing commits, only in two new commit).

File: tuf/api/metadata.py [hunk 7]
```diff
@@ -266,6 +269,63 @@ def sign(
...
+                role = next(
+                    (r for r in roles if r.name == delegated_role), None
+                )
+        else:
+            raise TypeError("Call is valid only on delegator metadata")
+
+        if role is None:
+            raise ValueError(f"No delegation found for {delegated_role}")
+
+        # verify that delegate is signed by required threshold of unique keys
```
Comment (joshuagl) [inline tuf/api/metadata.py]: Nit:
```suggestion
        # verify that delegated_metadata is signed by required threshold of unique keys
```

File: tuf/api/metadata.py [hunk 8]
```diff
@@ -266,6 +269,63 @@ def sign(
...
+                threshold of keys for 'role_name'
+        """
+
+        # Find the keys and role in delegator metadata
+        role = None
+        if isinstance(self.signed, Root):
+            keys = self.signed.keys
+            role = self.signed.roles.get(delegated_role)
+        elif isinstance(self.signed, Targets):
+            if self.signed.delegations:
```
Comment (joshuagl) [inline tuf/api/metadata.py]: Nit: I wonder if this would be slightly clearer if we did something like:
```Python
if not self.signed.delegations:
    raise TypeError("Targets does not delegate to any other roles")
keys = ...
...
```
?

File: tuf/api/metadata.py [hunk 9]
```diff
@@ -266,6 +269,63 @@ def sign(
...
         return signature
 
+    def verify_delegate(
+        self,
+        delegated_role: str,
+        delegated_metadata: "Metadata",
+        signed_serializer: Optional[SignedSerializer] = None,
+    ) -> None:
+        """Verifies that 'delegate' is signed with the required threshold of
+        keys for the delegated role 'role_name'.
```
Comment (joshuagl) [inline tuf/api/metadata.py]:
```suggestion
        """Verifies that 'delegated_metadata' is signed with the required
        threshold of keys for the 'delegated_role'.
```

**Review by joshuagl — APPROVED** [review]
I really like how easy this is to follow, nice work.

**Comment by jku** [comment]
Yes, agree with all suggestions. I've combined all the improvements from suggestions into one commit.

The changes since last push are
* use `if self.signed.delegations is None` to make that path clearer as suggested
* Add a test for this specific case (this required modifying the test data)
* tweak comments as suggested

It's minor enough that I don't think this absolutely needs another approval (so I will merge with the one I have) but ... there's not a lot of code that's more critical for TUF so feel free to have another look

**Review by MVrachev — APPROVED** [review]
I looked at the new changes and it seems good to me!

**Review by joshuagl — APPROVED** [review]
LGTM, thanks!
